### PR TITLE
Project Management Automation: Support adding milestones for fork PRs.

### PR DESCRIPTION
**Pull request by epiqueras**
Follows #20049.

>Separately, there might be alternatives to explore here, where the event is handled in response to the commit as it ends up being pushed to master. In other words, some combination of:
>
>- The push event
>- Conditioning on commits that occur in the master branch
>- Deriving a pull request associated with the commit
>- Applying milestone to the pull request

### CODE REVIEW

File: .github/workflows/pull-request-automation.yml [hunk 1]
```diff
@@ -1,6 +1,7 @@
 on:
   pull_request:
     types: [opened, closed]
```
Comment (aduth) [inline .github/workflows/pull-request-automation.yml]: We could revert the changes from #19742, since the only task which relied on `closed` is now using a different event.

File: packages/project-management-automation/lib/add-milestone.js [hunk 2]
```diff
@@ -32,19 +32,18 @@ const isDuplicateValidationError = ( error ) =>
...
+	if ( payload.ref !== 'refs/heads/master' ) {
+		debug( 'add-milestone: Commit is not to `master`. Aborting' );
 		return;
 	}
 
-	if ( payload.pull_request.base.ref !== 'master' ) {
-		debug(
-			'add-milestone: Pull request is not based on `master`. Aborting'
-		);
+	const [ , prNumber ] = payload.commits[ 0 ].message.match( /\(#(\d+)\)$/ );
```
Comment (aduth) [inline packages/project-management-automation/lib/add-milestone.js]: Will it work to match the parenthesized pull request number as always at the end of the message?

For example, looking at the webhook payload for 818eb80edccbba5445e0eadeaf9c440e22fe6a86, I see:

```json5
{
  "commits": [
    {
      "message": "Fix plain text selection/focus (#19964)\n\n* Upgrade react native to 0.62.0-rc.1\r\n\r\n* Fix focus on plain text for RN\r\n\r\nIt looks there was a type and we were checking the incorrect props.\r\n\r\n* Revert change on package.json\r\n\r\nCo-authored-by: Marko Savic <savicmarko1985@gmail.com>",
      // ...
    }
  ],
  // ...
}
```

```js
"Fix plain text selection/focus (#19964)\n\n* Upgrade react native to 0.62.0-rc.1\r\n\r\n* Fix focus on plain text for RN\r\n\r\nIt looks there was a type and we were checking the incorrect props.\r\n\r\n* Revert change on package.json\r\n\r\nCo-authored-by: Marko Savic <savicmarko1985@gmail.com>".match( /\(#(\d+)\)$/ )
// null
```

I think it might be enough to use the `m` (multiline) flag:

```suggestion
	const [ , prNumber ] = payload.commits[ 0 ].message.match( /\(#(\d+)\)$/m );
```

File: packages/project-management-automation/lib/index.js [hunk 3]
```diff
@@ -25,8 +25,7 @@ const automations = [
 		task: ifNotFork( addFirstTimeContributorLabel ),
 	},
 	{
-		event: 'pull_request',
-		action: 'closed',
+		event: 'push',
 		task: ifNotFork( addMilestone ),
```
Comment (aduth) [inline packages/project-management-automation/lib/index.js]: Is `ifNotFork` necessary anymore?

**Review by aduth — CHANGES_REQUESTED** [review]

File: .github/workflows/pull-request-automation.yml (hunk 1 above)
Comment (epiqueras) [reply .github/workflows/pull-request-automation.yml]: Done

File: packages/project-management-automation/lib/add-milestone.js (hunk 2 above)
Comment (epiqueras) [reply packages/project-management-automation/lib/add-milestone.js]: Good catch! Descriptions are concatenated there.

File: packages/project-management-automation/lib/index.js (hunk 3 above)
Comment (epiqueras) [reply packages/project-management-automation/lib/index.js]: Yes, why wouldn't it be?

Comment (aduth) [reply packages/project-management-automation/lib/index.js]: It doesn't seem we would ever have a scenario where we would receive a `push` event and also not immediately bail at the [first condition checking the `master` branch](https://github.com/WordPress/gutenberg/pull/20058/files#diff-c856bdd8bfe0b3c6917df8184fbab9adR39) (i.e. I can't see how an error would ever occur if the `ifNotFork` was not here). I suppose it might still be better to stop it as early as possible to avoid tying it to this specific implementation.

**Review by aduth — APPROVED** [review]

File: packages/project-management-automation/lib/index.js (hunk 3 above)
Comment (epiqueras) [reply packages/project-management-automation/lib/index.js]: Forks can also have branches named `master`.

Comment (aduth) [reply packages/project-management-automation/lib/index.js]: Are we receiving `push` events for those?

Comment (epiqueras) [reply packages/project-management-automation/lib/index.js]: Yes, someone could make a PR from `fork/master`.

Comment (aduth) [reply packages/project-management-automation/lib/index.js]: Hm, ok. I think I may have been expecting that [this condition](https://github.com/WordPress/gutenberg/pull/20058/files#diff-c856bdd8bfe0b3c6917df8184fbab9adR39-R42) would somehow already be verifying that the commit happened on `wordpress/gutenberg`'s `master`, but I see now that's not the case.

But! `ifNotFork` was implemented for pull request events, and I think would need to be updated to support this.

https://github.com/WordPress/gutenberg/blob/cf1a6d06b0f047c5942af6190d9789b4509323f8/packages/project-management-automation/lib/if-not-fork.js#L18-L19

Notably, `payload.pull_request` doesn't exist for `push` events.

https://developer.github.com/v3/activity/events/types/#pushevent

**Review by aduth — CHANGES_REQUESTED** [review]
https://github.com/WordPress/gutenberg/pull/20058#discussion_r375608937

File: packages/project-management-automation/lib/index.js (hunk 3 above)
Comment (epiqueras) [reply packages/project-management-automation/lib/index.js]: Then it's better to just not use it. I've removed it. No one actually makes PRs from `master` anyway. We can live without milestones for those edge cases.

**Comment by epiqueras** [comment]
https://github.com/WordPress/gutenberg/pull/20058#discussion_r375612215

File: packages/project-management-automation/lib/index.js (hunk 3 above)
Comment (aduth) [reply packages/project-management-automation/lib/index.js]: > Then it's better to just not use it. I've removed it. No one actually makes PRs from `master` anyway. We can live without milestones for those edge cases.

I don't know how to search for specific examples to share, but I've definitely seen pulls from forked `master` before. That said, the combination of this _and_ a commit message which matches the `(#PR)` pattern to get to the point of adding the milestone would be exceedingly rare, yes. And the result would merely be a failing action, nothing too worrisome.

We might want to revisit this in the future to make the logic more durable and to support `ifNotFork` usage for non-`pull_request` events (especially given the fact the name is generic enough to easily misinterpret that it ought be supported), but I don't think it's a blocker.

**Review by aduth — APPROVED** [review]

**Comment by aduth** [comment]
Small issue with the implementation which causes errors for commits to master: #20144

**Comment by epiqueras** [comment]
Good catch! Fixed: #20147.
